src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs

```diff
@@ -26,77 +27,85 @@ public static class ConcurrencyUtilities
+                            fs = AcquireFileStream(lockPath);
+                        }
+                        catch (DirectoryNotFoundException)
+                        {
+                            throw;
+                        }
+                        catch (PathTooLongException)
+                        {
+                            throw;
+                        }
+                        catch (UnauthorizedAccessException)
                         {
-                            if (string.IsNullOrEmpty(lockPath))
+                            token.ThrowIfCancellationRequested();
+
+                            if (unauthorizedAttemptsLeft < 1)
                             {
-                                lockPath = BasePath;
+                                if (string.IsNullOrEmpty(lockPath))
+                                {
+                                    lockPath = BasePath;
+                                }
+
+                                var message = string.Format(
+                                    CultureInfo.CurrentCulture,
+                                    Strings.UnauthorizedLockFail,
+                                    lockPath,
+                                    filePath);
+
+                                throw new InvalidOperationException(message);
                             }
 
-                            var message = string.Format(
-                                CultureInfo.CurrentCulture,
-                                Strings.UnauthorizedLockFail,
-                                lockPath,
-                                filePath);
+                            unauthorizedAttemptsLeft--;
 
-                            throw new InvalidOperationException(message);
+                            // This can occur when the file is being deleted
+                            // Or when an admin user has locked the file
+                            await Task.Delay(SleepDuration);
+                            continue;
                         }
+                        catch (IOException)
+                        {
+                            token.ThrowIfCancellationRequested();
 
-                        unauthorizedAttemptsLeft--;
-
-                        // This can occur when the file is being deleted
-                        // Or when an admin user has locked the file
-                        await Task.Delay(SleepDuration);
-                        continue;
-                    }
-                    catch (IOException)
-                    {
-                        token.ThrowIfCancellationRequested();
+                            await Task.Delay(SleepDuration);
+                            continue;
+                        }
 
-                        await Task.Delay(SleepDuration);
-                        continue;
+                        // Run the action within the lock
+                        return await action(token);
```

Comment:
I think we should consider some refactoring because code inside `while` loop looks almost identical in `ExecuteWithFileLockedAsync` and `ExecuteWithFileLocked` methods.

Reply:
If you check the PR using "ignore whitespace" (add `?w=1` to the url), you'll see the method is unchanged, just indented one level. I consider it outside the scope of this PR to refactor the methods. I suggest creating a new issue if you feel strongly about it.

Reply:
I had a quick look at the methods and the only thing that I could see being possible to refactor is the unauthorised exception handler. Calling the async version from the sync method and using `.Result` or `.GetAwaiter().GetResult()` is incorrect, as it can lead to deadlocks when on a thread with a synchronisation context. Calling the sync version from the async version is invalid because it uses Thread.Sleep and therefore blocks the thread.

Perhaps I'm not creative enough, but I don't see how this could be refactored to remove any more duplication in a meaningful way. It's just a natural consequence of having both a sync and async version of the method and doing it properly.